pkg/datapath/iptables/iptables.go

```diff
@@ -929,147 +1084,39 @@ func (m *IptablesManager) InstallRules(ifName string) error {
-			// are considered.
-			// Exclude traffic for other than interface from the masquarade rules.
-			// RETURN fro the chain as it is possible that other rules need to be matched.
-			if err := runProg("iptables", append(
-				m.waitArgs,
-				"-t", "nat",
-				"-A", ciliumPostNatChain,
-				"!", "-o", localDeliveryInterface,
-				"-m", "comment", "--comment", "exclude non-"+ifName+" traffic from masquerade",
-				"-j", "RETURN"), false); err != nil {
-				return err
-			}
-
-			// Exclude proxy return traffic from the masquarade rules
-			if err := runProg("iptables", append(
-				m.waitArgs,
-				"-t", "nat",
-				"-A", ciliumPostNatChain,
-				"-m", "mark", "--mark", matchFromProxy, // Don't match proxy (return) traffic
-				"-m", "comment", "--comment", "exclude proxy return traffic from masquarade",
-				"-j", "ACCEPT"), false); err != nil {
-				return err
-			}
-
-			if option.Config.Tunnel != option.TunnelDisabled {
-				// Masquerade all traffic from the host into the ifName
-				// interface if the source is not the internal IP
-				//
-				// The following conditions must be met:
-				// * Must be targeted for the ifName interface
-				// * Must be targeted to an IP that is not local
-				// * Tunnel mode:
-				//   * May not already be originating from the masquerade IP
-				// * Non-tunnel mode:
-				//   * May not orignate from any IP inside of the cluster range
-				if err := runProg("iptables", append(
-					m.waitArgs,
-					"-t", "nat",
-					"-A", ciliumPostNatChain,
-					"!", "-s", node.GetHostMasqueradeIPv4().String(),
-					"!", "-d", node.GetIPv4AllocRange().String(),
-					"-o", "cilium_host",
-					"-m", "comment", "--comment", "cilium host->cluster masquerade",
-					"-j", "SNAT", "--to-source", node.GetHostMasqueradeIPv4().String()), false); err != nil {
-					return err
-				}
-			}
+		if err := m.installMasqueradeRules("iptables", ifName, localDeliveryInterface,
+			datapath.RemoteSNATDstAddrExclusionCIDRv4().String(),
+			node.GetIPv4AllocRange().String(),
+			node.GetHostMasqueradeIPv4().String(),
+		); err != nil {
+			return err
+		}
+	}
 
-			// Masquerade all traffic from the host into local
-			// endpoints if the source is 127.0.0.1. This is
-			// required to force replies out of the endpoint's
-			// network namespace.
-			//
-			// The following conditions must be met:
-			// * Must be targeted for local endpoint
-			// * Must be from 127.0.0.1
-			if err := runProg("iptables", append(
-				m.waitArgs,
-				"-t", "nat",
-				"-A", ciliumPostNatChain,
-				"-s", "127.0.0.1",
-				"-o", localDeliveryInterface,
-				"-m", "comment", "--comment", "cilium host->cluster from 127.0.0.1 masquerade",
-				"-j", "SNAT", "--to-source", node.GetHostMasqueradeIPv4().String()), false); err != nil {
-				return err
-			}
+	if option.Config.EnableIPv6 {
+		// See comment above.
+		matchFromProxy := fmt.Sprintf("%#08x/%#08x", linux_defaults.MagicMarkIsProxy, linux_defaults.MagicMarkProxyMask)
```

Comment:
Maybe we could create a function to install the MARK rule, and then call with with `"ip{,6}tables"` params?